packages/ember-glimmer/lib/renderer.js

```diff
@@ -40,6 +40,49 @@ class DynamicScope {
   }
 }
 
+class RootState {
+  constructor(root, env, template, self, parentElement, dynamicScope) {
+    assert(`You cannot render \`${self.value()}\` without a template.`, template);
+
+    this.id = getViewId(root);
+    this.root = root;
+    this.result = undefined;
+    this.shouldReflush = false;
+
+    let options = this.options = {
+      alwaysRevalidate: false
+    };
+
+    this.render = () => {
+      let result = this.result = template.asEntryPoint().render(self, env, {
+        appendTo: parentElement,
+        dynamicScope
+      });
+
+      // override .render function after initial render
+      this.render = () => {
+        result.rerender(options);
+      };
+    };
```

Comment:
It seems like we can make all of this a prototype method by storing all these closed over states

```
render() {
  if (this.result) {
    // ...
  } else {
    // ...
  }
}
```